### ISSUE TITLE
Leverage vscode "Piece Tree" string buffer in TextDocument representation

### ISSUE THREAD

**Issue by Jma353** [microsoft/vscode-languageserver-node#476]
@rebornix recently made a [post](https://code.visualstudio.com/blogs/2018/03/23/text-buffer-reimplementation) w.r.t. text buffering in the main `vscode` project.  See [this code](https://github.com/Microsoft/vscode/blob/314cc521d8ce0c55773850ef520095afde33a62c/src/vs/editor/common/model/pieceTreeTextBuffer/pieceTreeTextBuffer.ts).

It would be awesome if `FullTextDocument`, `TextDocument` implementation used by things such `TextDocuments` (document manager bundled in the `server` package), used this buffering under the hood, rather than a plain string / lazy-loaded offset mapping.  

This buffered approach, combined with a supporting incremental sync in `TextDocuments` API (something I want to put a PR up for, and mentioned in #473) would afford language server / client developers a much more performant API for dealing with text documents.

Could we move this text buffer implementation into this repo?  Make it a separate npm module?  Open to peoples' thoughts on this?

cc: @dbaeumer @jrieken

**Comment by Jma353** [comment]
Additionally, totally understand the dependencies the vscode piece tree implementation has on other code that is exclusively in the vscode repo.  I guess I hoped my creation of this issue would also spur a discussion of efficient ways of buffering document text in text document implementations in this repo, as they are the default implementations developers inherit when leveraging the tooling in this package to create language servers.

**Comment by rebornix** [comment]
Technically we can move the [`pieceTreeBase`](https://github.com/Microsoft/vscode/blob/314cc521d8ce0c55773850ef520095afde33a62c/src/vs/editor/common/model/pieceTreeTextBuffer/pieceTreeBase.ts) code into a separate npm module then Monaco or vscode-languageserver-node can have its own implementation of `applyEdits` (Monaco's a bit more complex as it handles markers and whitespaces trimming). The only code change I need to make is moving searching related code out as all other vscode code dependencies are just types.

Moreover, we can leverage [`TextBuferSnapshot`](https://github.com/Microsoft/vscode/blob/314cc521d8ce0c55773850ef520095afde33a62c/src/vs/editor/common/model/pieceTreeTextBuffer/pieceTreeBase.ts#L159) for snapshot and incremental update.

**Comment by Jma353** [comment]
@rebornix that sounds great.  Is this something that would take a great deal of time?  Is this something you're interested in doing?

Happy to help where I can.

**Comment by rebornix** [comment]
Publishing `PieceTreeBase` should be simple. 

It doesn't mean that we have to use that in vscode-languageserver-node as a line based text buffer can work great for small to medium files. The discussion came back to If we want to support incremental content sync in the document manager, as mentioned in https://github.com/Microsoft/vscode-languageserver-node/issues/473#issuecomment-476549177 .

**Comment by Jma353** [comment]
By a line-based text buffer, do you mean an array of lines?  

As it stands today, I don't think this library uses a representation like that.  I am specifically looking at [`FullTextDocument`](https://github.com/Microsoft/vscode-languageserver-node/blob/9de8809400b7c443bc25f211dfaddcb74675b2d7/types/src/main.ts#L2614) and [`TextDocument.applyEdits`](https://github.com/Microsoft/vscode-languageserver-node/blob/9de8809400b7c443bc25f211dfaddcb74675b2d7/types/src/main.ts#L2488-L2510) (which actually uses repeated string concatenation. 

From what I can see, the document contents are always stored as a plain `string`. 

I plan on writing up a PR to add Incremental document sync support to `TextDocuments`, leveraging things like `TextDocument.applyEdits` to achieve a logically correct synchronization mechanism.  I guess I wanted to kick off this discussion as the actual efficiency of `applyEdits` could be improved.

**Comment by dbaeumer** [comment]
@Jma353 a PR is highly appreciated.

**Comment by rebornix** [comment]
@Jma353 I extracted PieceTree from vscode to a separate repository https://github.com/rebornix/PieceTree, so you can play with it for now for the incremental feature. You may also want to read how we handle [`applyEdits`](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/common/model/pieceTreeTextBuffer/pieceTreeTextBuffer.ts#L174) in vscode.

Once you have something working (say a PR), we can look into publishing PieceTree to NPM and write some scripts to automate the publishing (we probably don't need to as PieceTree is stable).

**Comment by Jma353** [comment]
Thanks for the super-fast turnaround time @rebornix!  

I'll be putting out a PR next week!

**Comment by Jma353** [comment]
@rebornix thanks again for separating out your tree-based text buffer.

Here's how I'm envisioning we introduce this: 
1. I introduce a PR with incremental doc sync added to `TextDocuments` document manager, using functions like `TextDocument.update` and `TextDocument.applyEdits` as black-boxes.  These functions will be swapped out for more efficient variants in the future, using underlying piece-tree text buffer. 
2. You publish https://github.com/rebornix/PieceTree as an NPM module.
3. You or me put out a PR adding that npm module as a dependency of this project.
4. I swap out the `TextDocument.applyEdits` function and `FullTextDocument` implementations for ones that are more efficient, and use the tree-based text buffer in a PR.  Once this is in place, incremental doc-sync with efficient text buffering will be available out-of-the-box when using `TextDocuments` doc manager in an LSP server. 

I am working on 1. from this list right now.  Does this plan sound good to you?

**Comment by Jma353** [comment]
#478

**Comment by rebornix** [comment]
> You publish https://github.com/rebornix/PieceTree as an NPM module

I published https://www.npmjs.com/package/vscode-piece-tree directly from the repository now. 

> 3. You or me put out a PR adding that npm module as a dependency of this project.
> 4. I swap out the TextDocument.applyEdits function and FullTextDocument implementations for ones that are more efficient, and use the tree-based text buffer in a PR. Once this is in place, incremental doc-sync with efficient text buffering will be available out-of-the-box when using TextDocuments doc manager in an LSP server.

I'm not sure if we want to separate the incremental document sync support into two PRs. If leveraging piece tree or a line based (an array of lines) text buffer improve the performance noticeably, we may want to do them directly but it's up to you and @dbaeumer to decide .

**Comment by Jma353** [comment]
I am happy to directly follow up #478 with a PR adding your npm module as a dependency + refactor the text modification logic of `FullTextDocument` and `TextDocument` `namespace`.  Or I can rebase #478 onto the PR adding and using your npm module. Whatever is easiest.

**Comment by rebornix** [comment]
Can you add the change (adding npm dependency, and refactor) directly to #478? I'm not confident with the performance of how `applyEdits` is being executed in the incremental sync. As I mentioned in https://github.com/Microsoft/vscode-languageserver-node/pull/478/files#r273749432 , every content change will lead to a re-scan of the whole document.

**Comment by dbaeumer** [comment]
@Jma353 and @rebornix thanks a lot for doing this. 

I would like to ask for one thing: can we still keep the old full text sync in and allow users to actively decide what they want to do. I am not a fan of swapping this out without server implementors actively deciding on it. Reason is that a full document sync works even if the client doesn't correctly provide character offsets in UTF-16 encoding. If we incrementally update a client needs to ensure this.

**Comment by dbaeumer** [comment]
I saw in https://github.com/Microsoft/vscode-languageserver-node/pull/478 that this is a constructor option which is good.

**Comment by Jma353** [comment]
@rebornix - I can definitely add the dependency in the PR!

Sounds good @dbaeumer, I'll keep this in mind.  Yea, figured I'd have the constructor option w/default to `Full` for that reason.

**Comment by Jma353** [comment]
@rebornix as a heads up, I got a `EISGIT` error when `npm install`-ing your module: 

```
npm ERR! code EISGIT
npm ERR! git <path/to/project/root>/node_modules/vscode-piece-tree: Appears to be a git repo or submodule.
npm ERR! git     <path/to/project/root>/node_modules/vscode-piece-tree
npm ERR! git Refusing to remove it. Update manually,
npm ERR! git or move it out of the way first.
```

I unblocked myself to perform `npm install` updates by deleting the `.git` folder in the installed `vscode-piece-tree` module.  This might be something you need to update though, in order for this not to error out during `travis-ci` checks in the PR.

**Comment by rebornix** [comment]
@Jma353 the error should be gone with `vscode-piece-tree@1.1.1`.

**Comment by Jma353** [comment]
Updated the PR yesterday to use the Piece Tree, in a new `TextDocument` representation called `IncrementalTextDocument` (decided to go with that name, as there was a `FullTextDocument` built for `Full` text sync).  

Feel free to give it a look!

**Comment by Jma353** [comment]
Wanted to ping here to get additional feedback on my PR.  There's some comments w.r.t. updating documentation of parameters, but wanted to get additional feedback as well, so I can check this change in.

**Comment by dbaeumer** [comment]
Published server version `5.3.0-next.7`
